# Map, Filter, Reduce, Lambda & Recursion

## Tasks Today:

1) <b>Lambda Functions</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Saving to a Variable <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Multiple Inputs <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Passing a Lambda into a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; e) Returning a Lambda from a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; f) In-Class Exercise #1 <br>
2) <b>Map</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Map <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #2 <br>
3) <b>Filter</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Filter <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #3 <br>
4) <b>Reduce</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Reduce <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #4 <br>
5) <b>Recursion</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Implementing a Base <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Writing a Factorial Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #5 <br>
6) <b>Generators & Iterators</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Yield Keyword <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Inifinite Generator <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #6 <br>
7) <b>Exercises</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Exercise #1 - Filtering Empty Strings <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Exercise #2 - Sorting with Last Name <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Exercise #3 - Conversion to Farhenheit <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Exercise #4 - Fibonacci Sequence <br>

## Lambda Functions <br>
<p>Lambda functions... or "Anonymous Functions" are referring to inline functions with no name. The keyword lambda denotes the no name function, and executes within a single line. Without saving it to a variable; however, it is not able to be used, unless passed in either as a paramater or within list comprehension.<br>Written as "(keyword lambda) (one or more inputs) (colon) (function to be executed)"</p>

#### Syntax

In [1]:
def addTwo(x):
    return x+2

print(addTwo(3))

#as a lambda function 

lambda x : x+2

print((lambda x: x+2)(3))

5
5


#### Saving to a Variable

In [5]:
f = lambda x, y: x+y+2

print(f(2,3))
print((lambda x, y: x+y+2)(2,3))



7
7


#### Multiple Inputs

In [7]:
print((lambda x,y,z:x*y*z)(3,5,8))

x = lambda x,y,z:x*y*z
print(x(3,5,8))

120
120


#### Passing a Lambda into a Function

In [8]:
def multiply(f, x):
    return f(x)

multiply(lambda x: x*x, 7)

49

#### Returning a Lambda from a Function

In [11]:
#regular defined function 
def mupt(num):
    return num*4

#function within a function
def returnFunc(): 
    test = 4
    def mult(num):
        return num * 2
    return mult

#assign funtion to variable
f_return = returnFunc()

print(returnFunc())
print(f_return(4))

<function returnFunc.<locals>.mult at 0x000001339A842670>
8


In [13]:
#Lambda function returned from a regular function
def returnLam(b,c):
    return lambda x,a: x+a + b + c

#entering 4 and 5 as variables for the function returnLam and assigning it to a variable
#x and a are still undefined. this may somehow be connected to the init & methods from prev.
r_lam = returnLam(4,5)

print(r_lam)

r_lam(4,5)
# print(r_lam(4,5))

<function returnLam.<locals>.<lambda> at 0x000001339A842280>


18

#### If Statements within Lambdas

In [14]:
#Lambda x : True if (condition) else false
f = lambda n: n*2 if n>10 else n+2

f(12)
f(10)


12

#### In-Class Exercise #1 <br>
<p>Write an anonymous function that cubes the arguments passed in and assign the anonymous function to a variable 'f'.</p>

In [15]:
f = lambda x: x**3

## Map <br>
<p>The map function allows you to iterate over an entire list while running a function on each item of the list. This is why the map function works well with lambda's, because it simplifies things and you write less lines of code.<br>The syntax for a map function is "map(function to be used, list to be used)"<br>However, you must be careful, as the map function returns a map object, not a list. To turn it into a list we use the list() type conversion.</p>

#### Syntax

In [ ]:
# map(func,iterable(list,dict,tuple,set, etc))

In [16]:
def squared(num, num2):
    if num < 10 and num2 < 10:
        return num**2, num2**2
    else:
        return num, num2

num=[4,10,5,4,2,]
num2=[4,11,20,3,15]

squared_nums_map = map(squared, num, num2)
print(squared_nums_map)
print(list(squared_nums_map))

[(16, 16), (10, 11), (5, 20), (16, 9), (2, 15)]


#### Using Lambda's with Map

In [ ]:
# map(Lambda x:do something,iterable)
# using Lambda in map happens in line (or in one line)
# list(map(lambda x: x*2,nums))

In [18]:
squared = list(map(lambda x,y: (x**2,y**2) if x < 10 and y<10 else (x,y), num, num2))
print(squared)

[(16, 16), (10, 11), (5, 20), (16, 9), (2, 15)]


#### In-Class Exercise #2 <br>
<p>Use the map function to double each number and minus it by one in the list by using a lambda function</p>

In [24]:
nums = [4,2,6,8]

dub_minus_one = list(map(lambda x: x*2 - 1, nums))

print(dub_minus_one)

[7, 3, 11, 15]


## Filter() <br>
<p>Filter's are similar to the map function, where you're able to pass a function argument and a list argument and filter out something from the list based on the conditions passed. Similar to the map function, it returns a filter object, so you need to type convert it to a list()</p>

#### Syntax

In [25]:
names = ['Anna', 'Beth', 'David', 'Daniel', 'Kristina', 'Krystyna', 'David']

def d_name(name): 
    if name[0].lower() == 'd':
        return True
    else:
        return False
    
new_names = list(filter(d_name, names))
print(new_names)

['David', 'Daniel', 'David']


#### Using Lambda's with Filter()

In [26]:
new_names = list(filter(lambda name: True if name[0].lower() == 'd' else False, names))
print(new_names)

['David', 'Daniel', 'David']


#### In-Class Exercise #3 <br>
<p>Filter out all the numbers that are below the mean of the list.<br><b>Hint: Import the 'statistics' module</b></p>

In [27]:
from statistics import mean
num_list = [2,3,5,8]
new_num = list(filter(lambda num: True if num >= mean(num_list) else False, num_list))
print(new_num)

[5, 8]


## Reduce() <br>
<p>Be very careful when using this function, as of Python 3 it's been moved to the 'functools' library and no longer is a built-in function.<br>The creator of Python himself, says to just use a for loop instead.</p>

#### Syntax

In [28]:
from functools import reduce
# reduce(function,iterable)
help(reduce)

Help on built-in function reduce in module _functools:

reduce(...)
    reduce(function, sequence[, initial]) -> value
    
    Apply a function of two arguments cumulatively to the items of a sequence,
    from left to right, so as to reduce the sequence to a single value.
    For example, reduce(lambda x, y: x+y, [1, 2, 3, 4, 5]) calculates
    ((((1+2)+3)+4)+5).  If initial is present, it is placed before the items
    of the sequence in the calculation, and serves as a default when the
    sequence is empty.



In [31]:
list = [2,4,6,8,10]

def add(num1, num2):
    return num1 + num2

result = reduce(add, list)
print(result)

def sub(num1, num2):
    return num1 - num2

result = reduce(sub, list)
print(result)

30
-26


#### Using Lambda's with Reduce()

In [32]:
result = reduce(lambda x,y: x+y,list)
print(result)

30


#### In-Class Exercise #4 <br>
<p>Use the reduce function to multiply the numbers in the list below together with a lambda function.</p>

In [33]:
my_list = [1,2,3,4]

result = reduce(lambda x,y: x*y, my_list)
print(result)

24


## Recursion <br>
<p>Recursion means that a function is calling itself, so it contanstly executes until a base case is reached. It will then push the returning values back up the chain until the function is complete. A prime example of recursion is computing factorials... such that 5! (factorial) is 5*4*3*2*1 which equals 120.</p>

#### Implementing a Base Case

In [41]:
def addNums(num):
    if num <=1:
        print("addNums(1) = 1")
        return num
    else:
        print(f'addNums({num}) = {num} + addNums({num-1})')
        
addNums(5)
print('---------------------------')

def addNums(num):
    if num <=1:
        print("addNums(1) = 1")
        return num
    else:
        print(f'addNums({num}) = {num} + addNums({num-1})')
        return num + addNums(num-1)

addNums(5)

addNums(5) = 5 + addNums(4)
---------------------------
addNums(5) = 5 + addNums(4)
addNums(4) = 4 + addNums(3)
addNums(3) = 3 + addNums(2)
addNums(2) = 2 + addNums(1)
addNums(1) = 1


15

#### Writing a Factorial Function

In [42]:
def fact(num):
    if num <= 1:
        return 1
    else:
        return num*fact(num-1)
fact(5)

120

#### In-Class Exercise #5 <br>
<p>Write a recursive function that subtracts all numbers to the argument given.</p>

In [43]:
def rec_sub(num):
    if num <= 1:
        return(1)
    else:
        return num - rec_sub(num-1)
    
rec_sub(10)

5

## Generators <br>
<p>Generators are a type of iterable, like lists or tuples. They do not allow indexing, but they can still be iterated through with for loops. They are created using functions and the yield statement.</p>

#### Yield Keyword <br>
<p>The yield keyword denotes a generator, it doesn't return so it won't leave the function and reset all variables in the function scope, instead it yields the number back to the caller.</p>

In [47]:
def my_range(stop, start, step = 2):
    while start<stop:
        yield start
        start += step
        
for i in my_range(20, start = 2):
    my_gen_value = i
    print(my_gen_value)

2
4
6
8
10
12
14
16
18


#### Infinite Generator

In [ ]:
# bad, never create infinite loops

#### In-Class Exercise #6 <br>
<p>Create a generator that takes a number argument and yields that number squared, then prints each number squared until zero is reached.</p>

In [48]:
def square(stop, start, step=1):
    while start>stop:
        yield start**2
        start -= step 
    
for num in square(0,10,1):
    print(num)

100
81
64
49
36
25
16
9
4
1


# Exercises

### Exercise #1 <br>
<p>Filter out all of the empty strings from the list below</p>

`Output: ['Argentina', 'San Diego', 'Boston', 'New York']`

In [2]:
places = [" ","Argentina", " ", "San Diego","","  ","","Boston","New York"]

places_filtered = list(filter(lambda place: False if place.strip() == "" else True, places))

print(places_filtered)

['Argentina', 'San Diego', 'Boston', 'New York']


### Exercise #2 <br>
<p>Write an anonymous function that sorts this list by the last name...<br><b>Hint: Use the ".sort()" method and access the key"</b></p>

`Output: ['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']`

In [3]:
author = ["Joel Carter", "Victor aNisimov", "Andrew P. Garfield","David hassELHOFF","Gary A.J. Bernstein"]

author.sort(key= lambda name: name.lower().split()[-1])

print(author)

['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']


### Exercise #3 <br>
<p>Convert the list below from Celsius to Farhenheit, using the map function with a lambda...</p>

`Output: [('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]
`

In [4]:
# F = (9/5)*C + 32
places = [('Nashua',32),("Boston",12),("Los Angelos",44),("Miami",29)]

list(map(lambda place: [place[0], place[1]*(9/5) + 32], places))

[['Nashua', 89.6], ['Boston', 53.6], ['Los Angelos', 111.2], ['Miami', 84.2]]

### Exercise #4 <br>
<p>Write a recursion function to perform the fibonacci sequence up to the number passed in.</p>

`Output for fib(5) => 
Iteration 0: 1
Iteration 1: 1
Iteration 2: 2
Iteration 3: 3
Iteration 4: 5
Iteration 5: 8`

In [28]:
def fib(num):
    if num <= 1:
        return 1
    else:
        return fib(num-1) + fib(num-2)

num=5
    
for i in range(num+1):
    print(f'Iteration {i}: {fib(i)}')

# print(fib(0))
# print(fib(1))
# print(fib(2))
# print(fib(3))
# print(fib(4))
# print(fib(5))

Iteration 0: 1
Iteration 1: 1
Iteration 2: 2
Iteration 3: 3
Iteration 4: 5
Iteration 5: 8
